types/wlr_xwayland_keyboard_grab_v1.c

```diff
@@ -0,0 +1,200 @@
+#include <stdlib.h>
+#include <assert.h>
+#include <wayland-server.h>
+#include <wayland-util.h>
+#include <wlr/util/log.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/types/wlr_xwayland_keyboard_grab_v1.h>
+
+#include <xwayland-keyboard-grab-unstable-v1-protocol.h>
+
+struct wlr_xwayland_keyboard_grab_v1_manager {
+	struct wl_resource *resource;
+	struct wlr_xwayland_keyboard_grab_v1 *xwayland_keyboard_grab;
+
+	struct wl_list link; // wlr_xwayland_keyboard_grab_v1::clients;
+	struct wl_list grabs;
+
+	struct {
+		struct wl_signal destroy;
+	} events;
+};
+
+static void xwayland_keyboard_grab_grab_handle_destroy(struct wl_resource *resource) {
+	assert(resource);
+	struct wlr_xwayland_keyboard_grab_v1_grab *grab =
+		wl_resource_get_user_data(resource);
```

Comment:
We now use `wl_resource_get_user_data` wrappers to ensure type safety. Have a look at other protocol implementations to see how to do it.